v2/car.go

```diff
@@ -41,12 +45,70 @@ type (
 	}
 )
 
-// Size gets the size of Header in number of bytes.
-func (h *Header) Size() int {
-	return HeaderBytesSize
+// WriteTo writes this characteristics to the given writer.
+func (c Characteristics) WriteTo(w io.Writer) (n int64, err error) {
+	if err = writeUint64To(w, c.Hi); err != nil {
+		return
+	}
+	n += uint64BytesSize
+	if err = writeUint64To(w, c.Lo); err != nil {
+		return
+	}
+	n += uint64BytesSize
+	return
+}
+
+// NewHeader instantiates a new CAR v2 header, given the byte length of a CAR v1.
+func NewHeader(carV1Size uint64) Header {
+	header := Header{
+		CarV1Size: carV1Size,
+	}
+	header.CarV1Offset = prefixBytesSize + HeaderBytesSize
+	header.IndexOffset = header.CarV1Offset + carV1Size
+	return header
+}
+
+// WithIndexPadding sets the index offset from the beginning of the file for this header and returns the
+// header for convenient chained calls.
+// The index offset is calculated as the sum of PrefixBytesLen, HeaderBytesLen,
+// Header.CarV1Len, and the given padding.
+func (h Header) WithIndexPadding(padding uint64) Header {
+	h.IndexOffset = h.IndexOffset + padding
```

Comment:
@masih when we do this, do we lose the ability to tell exactly where the CARv1 ends? I think we need to retain the ability to derive `carV1Size` so we can parse it successfully using the CARv1 parser, which only knows "end" when the reader runs out of bytes. If we feed it zero-padding at the end we're going to have problems I think--not just parsing problems either, if we continue to use exact CARv1 bytes to generate CommP for Filecoin then we need to know the exact bytes.

Reply:
fwiw re scale, Filecoin storage deals can be packed into ~30Gb CAR files, which is a common case we're going to need to be OK with (double that for miners doing 64G sectors). The [full chain snapshot for Filecoin](https://docs.filecoin.io/get-started/lotus/chain/#full-chain-snapshot) is probably the biggest CAR around, I think it's in the 420Gb range. The latter is probably not a CARv2 candidate, but the former is, at least on the miner side. So we're probably going to want a clear route to non-buffering writes.

Reply:
> when we do this, do we lose the ability to tell exactly where the CARv1 ends?

The header contains a dedicated [field for storing the size of car v1 dump](https://github.com/masih/go-car/blob/ad8a17eaac3171c30f9bfcadc6464bd8e5c96a3b/v2/car.go#L37). Did I miss something?

> a clear route to non-buffering writes

Of course; all of this is living in a WIP branch. No buffering of car content in memory for the final version :)

Reply:
> The header contains a dedicated field for storing the size of car v1 dump. Did I miss something?

nope, I did! all good then.